### DISCUSSION TITLE
ESQ1/SQ80 work around for [exit]

### DISCUSSION THREAD

**Discussion by troach242** [christofmuc/KnobKraft-orm#338] [General]
Awesome librarian. I sure wish I were a programmer and could help more.

When clicking on a program to send to my SQ80, the program is successfully set to the unit. Yay! The problem is the program gets stuck in the "write program" dialog and I need to press [exit]. This is very inconvenient as the synth is on the other side of the studio. PATCH BASE gets around this issue by sending a pre-sysex, the program sysex, and a post-sysex. How can I edit the adapation to add the pre and post sysex to the program?

Send this to the synth and the program is send and dialogs are avoided. 
pre = F0 0F 02 00 0E 26 59 22 55 F7
program = xxxxxxx
post = F0 0F 02 00 0E 2F 62 F7

**Comment by christofmuc** [comment]
Thank you! @Mostelin will be interested in this.

The SQ80 adaptation shares the code with the ESQ-1. Look for a file called "Ensoniqesq1.py" and copy it into a user folder. The default folder for KnobKraft user code can be set via the Options... Set User Adaptation Dir... menu, it points to a different place on each platform, so check the dialog to see where it expects user code. If you restart the KnobKraft, it will load your version of the adaptation first, confirming the override with a message in the log dialog. If you want to go back to the shipped adaptation, just delete the override file you just copied.

But when you have copied the file, you can try to add your additional messages to line 173 in the convertToEditBufferDump() function.

Currently, it returns the message with the channel changed. This could be amended by returning your messages like this:

```
def convertToEditBuffer(channel, message):
    # Instead of just returning the same message, this ensures that the most recently chosen channel is used.
    if isEditBufferDump(message):
        return [0xF0, 0x0F, 0x02, 0x00, 0x0E, 0x26, 0x59, 0x22, 0x55, 0xF7] + message[:3] + [channel] + message[4:] + [0xF0, 0x0F, 0x02, 0x00, 0x0E, 0x2F, 0x62, 0xF7]

```
There might be more involved especially as I also see a function for program dumps, but it is worth a try to see if that changes already something. You can use the MIDI log window to exactly see the bytes sent out by the KnobKraft to check that your change does what we wanted.

**Comment by Mostelin** [comment]
Yes, this certainly is interesting.

What @troach242 reports is exactly the intended behaviour of the current Ensoniq adaptation, because I was unaware of any way to EXIT input buffer editing other than manually. I share his feelings about this too; it is slightly clumsy.

I'll investigate his suggested SYSEX code in the next few days, and report back.

**Comment by troach242** [comment]
My SQ-80 challenge is solved. I added the pre & post commands to my adaptation as you instructed and poof... it works perfectly now. Thank you very much for the incredibly fast response and the amazing product.

**Comment by Mostelin** [comment]
I finally got around to looking at this. Sorry for the long delay; it was due to bronchitis.

Anyway, I've also made the changes discussed, and like how things work now. I've embedded the new code in the 0.92 version of the ESQ-1 adaptation and, to avoid confusion, modified the user messages that appear on the Adaptation tab - the warning about clearing the buffer has been removed.

I need to remind myself about the submission process for adaptation revisions...

Most Ensoniq users should be able to save newly downloaded sounds to any location in their program banks without further explanation. Just to make sure though, I think once the new adaptation is included in the download, I'll add an explanatory note in the Wiki.

Thanks to @troach242 and @christofmuc for this nice enhancement.

**Comment by Mostelin** [comment]
Trying to upload new version 0.92 here...
[Ensoniqesq1.92.zip](https://github.com/user-attachments/files/18032973/Ensoniqesq1.92.zip)

**Reply by christofmuc** [reply]
Great, I'll include it in the next build! Any release note sentence you I should use?

**Reply by Mostelin** [reply]
I've modified the adaptation notes that appear in ORM, and added something about this to the Wiki too. You could probably use the Wiki notes.

**Comment by Mostelin** [comment]
I have discovered a rather amusing feature. It is possible to load and hold multiple programs cumulatively. Click a program in ORM and press a key on the synth, click another program and press a different key without letting the first key up. Continue in this fashion until you run out of fingers. Works best with programs that sustain.

**Reply by christofmuc** [reply]
This blows my mind. Why does this work?

**Reply by Mostelin** [reply]
I have no idea why it works, and I cannot recall this ever being reported. I knew you can play multiple normal programs simultaneously, by changing programs between key presses, but how is this possible with programs that have not yet been saved? The thing is nearly 40 years old and still surprises me!

**Comment by christofmuc** [comment]
Closing this as the release 2.5.0 contains the new version from @Mostelin.
